### ISSUE TITLE
多元线性回归

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#12303]
多元线性回归<br>
<br>
https://ift.tt/9XATf8R<br>
<br>
<br>
<br>

<div><div>
<div>
                            <h2 id="toc-0">一. Multiple Features</h2>
<p>具有多个变量的线性回归也被称为“多元线性回归”。</p>
<p><span></span><span id="MathJax-Element-1-Frame" tabindex="0"><span><math>xj(i)</math></span></span>: 训练集第 i 个向量中的第 j 个元素(第 i 行第 j 列)<br><span></span><span id="MathJax-Element-2-Frame" tabindex="0"><span><math>x(i)</math></span></span>: 训练集第 i 个向量(第 i 行)<br><span></span><span id="MathJax-Element-3-Frame" tabindex="0"><span><math>m</math></span></span>: 总共 m 行<br><span></span><span id="MathJax-Element-4-Frame" tabindex="0"><span><math>n</math></span></span>: 总共 n 列</p>
<p>适应这些多特征的假设函数的多变量形式如下：</p>
<p>
<span></span><span><span id="MathJax-Element-5-Frame" tabindex="0"><span><math>hθ(x)=θ0+θ1x1+θ2x2+θ3x3+⋯+θnxn</math></span></span></span>
</p>
<p>使用矩阵乘法的定义，我们的多变量假设函数可以简洁地表示为：</p>
<p>
<span></span><span><span id="MathJax-Element-6-Frame" tabindex="0"><span><math>hθ(x)=[θ0θ1⋯θn][x0x1⋮xn]=θTx</math></span></span></span>
</p>
<p>其中 <span></span><span id="MathJax-Element-7-Frame" tabindex="0"><span><math>x0(i)=1(i∈1,⋯,m)</math></span></span></p>
<hr>
<h2 id="toc-1">二. Gradient Descent for Multiple Variables</h2>
<p>多个变量的梯度下降，同时更新 n 个变量。</p>
<p><span></span><span><span id="MathJax-Element-8-Frame" tabindex="0"><span><math>θj:=θj−α1m∑i=1m(hθ(x(i))−y(i))xj(i)</math></span></span></span></p>
<p>其中 <span></span><span id="MathJax-Element-9-Frame" tabindex="0"><span><math>j∈[0,n]</math></span></span></p>
<hr>
<h2 id="toc-2">三. Gradient Descent in Practice I - Feature Scaling</h2>
<p>特征缩放包括将输入值除以输入变量的范围（即最大值减去最小值），导致新的范围仅为1。</p>
<p>均值归一化包括从输入变量的值中减去输入变量的平均值，从而导致输入变量的新平均值为零。</p>
<h3 id="toc-3">1. Feature Scaling</h3>
<p>特征缩放让特征值取值范围都比较一致，这样在执行梯度下降的时候，“下山的路线”会更加简单，更快的收敛。通常进行特征缩放都会把特征值缩尽量缩放到 [-1,1] 之间<strong>或者这个区间附近</strong>。</p>
<p>即 <span></span><span id="MathJax-Element-10-Frame" tabindex="0"><span><math>xi=xisi</math></span></span></p>
<h3 id="toc-4">2. Mean normalization</h3>
<p><span></span><span id="MathJax-Element-11-Frame" tabindex="0"><span><math>xi=xi−μisi</math></span></span></p>
<p>其中，<span></span><span id="MathJax-Element-12-Frame" tabindex="0"><span><math>μi</math></span></span> 是特征值的所有值的平均值，<span></span><span id="MathJax-Element-13-Frame" tabindex="0"><span><math>si</math></span></span> 是值的范围（最大 - 最小），或者 <span></span><span id="MathJax-Element-14-Frame" tabindex="0"><span><math>si</math></span></span> 是标准偏差</p>
<p>当然 <span></span><span id="MathJax-Element-15-Frame" tabindex="0"><span><math>x0=1</math></span></span> 就不需要经过上述的处理了，因为它永远等于1，不能有均值等于0的情况。</p>
<hr>
<h2 id="toc-5">四. Gradient Descent in Practice II - Learning Rate</h2>
<p>如果学习率 <span></span><span id="MathJax-Element-16-Frame" tabindex="0"><span><math>α</math></span></span> 太小的话，就会导致收敛速度过慢的问题。<br>
如果学习率 <span></span><span id="MathJax-Element-17-Frame" tabindex="0"><span><math>α</math></span></span> 太大的话，代价函数可能不会在每次迭代中都下降，甚至可能不收敛，在某种情况下，学习率 <span></span><span id="MathJax-Element-18-Frame" tabindex="0"><span><math>α</math></span></span> 过大，也有可能出现收敛缓慢。</p>
<p>可以通过绘制代价函数随迭代步数变化的曲线去调试这个问题。</p>
<p><span></span><span id="MathJax-Element-19-Frame" tabindex="0"><span><math>α</math></span></span> 的取值可以从 0.001，0.003，0.01，0.03，0.1，0.3，1 这几个值去尝试，选一个最优的。</p>
<hr>
<h2 id="toc-6">五. Features and Polynomial Regression</h2>
<p>可以通过改造特征值，例如合并2个特征，用 <span></span><span id="MathJax-Element-20-Frame" tabindex="0"><span><math>x3</math></span></span> 来表示 <span></span><span id="MathJax-Element-21-Frame" tabindex="0"><span><math>x1∗x2</math></span></span></p>
<p>在多项式回归中，针对 <span></span><span id="MathJax-Element-22-Frame" tabindex="0"><span><math>hθ(x)=θ0+θ1x1+θ2x12+θ3x13</math></span></span> ，我们可以令 <span></span><span id="MathJax-Element-23-Frame" tabindex="0"><span><math>x2=x12,x3=x13</math></span></span> 降低次数。</p>
<p>还可以考虑用根号的式子，例如选用  <span></span><span id="MathJax-Element-24-Frame" tabindex="0"><span><math>hθ(x)=θ0+θ1x1+θ2x</math></span></span></p>
<p>通过上述转换以后，需要记得用<strong>特征值缩放，均值归一化，调整学习速率的方式调整一下</strong>。</p>
<hr>
<blockquote>
<p>GitHub Repo：<a href="https://github.com/halfrost/Halfrost-Field">Halfrost-Field</a></p>
<p>Follow: <a href="https://github.com/halfrost">halfrost · GitHub</a></p>
<p>Source: <a href="https://github.com/halfrost/Halfrost-Field/blob/master/contents/Machine_Learning/Multivariate_Linear_Regression.ipynb">https://github.com/halfrost/Halfrost-Field/blob/master/contents/Machine_Learning/Multivariate_Linear_Regression.ipynb</a></p>
</blockquote>

                            <img src="https://img.halfrost.com/wechat-qr-code.png">
</div>
                        
                    </div></div><br>
<br>
<br>
<br>
<br>
<br>
via Halfrost's Field | 冰霜之地 https://halfrost.com<br>
<br>
October 11, 2025 at 09:45AM
